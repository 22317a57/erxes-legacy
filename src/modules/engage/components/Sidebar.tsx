import Box from 'modules/common/components/Box';
import CountsByTag from 'modules/common/components/CountsByTag';
import { __, router } from 'modules/common/utils';
import Wrapper from 'modules/layout/components/Wrapper';
import { SidebarCounter, SidebarList } from 'modules/layout/styles';
import { ITag } from 'modules/tags/types';
import React from 'react';
import { Link } from 'react-router-dom';
import { MESSAGE_KIND_FILTERS, statusFilters } from '../constants';

const { Section } = Wrapper.Sidebar;

type Props = {
  kindCounts: any;
  statusCounts: any;
  tagCounts: any;
  tags: ITag[];
  history?: any;
  toggleSection: (params: { name: string; isOpen: boolean }) => void;
  config: { [key: string]: boolean };
};

class Sidebar extends React.Component<Props> {
  renderKindFilter() {
    const { kindCounts, history } = this.props;

    return (
      <Section>
        <Section.Title>{__('Kind')}</Section.Title>

        <SidebarList>
          <li>
            <Link to="/engage">
              {__('All')}
              <SidebarCounter>{kindCounts.all}</SidebarCounter>
            </Link>
          </li>

          {MESSAGE_KIND_FILTERS.map((kind, index) => (
            <li key={index}>
              <Link
                tabIndex={0}
                className={
                  router.getParam(history, 'kind') === kind.name ? 'active' : ''
                }
                to={`/engage?kind=${kind.name}`}
              >
                {__(kind.text)}
                <SidebarCounter>{kindCounts[kind.name]}</SidebarCounter>
              </Link>
            </li>
          ))}
        </SidebarList>
      </Section>
    );
  }

  renderStatusFilter() {
    const { statusCounts, history } = this.props;

    return (
      <Section>
        <Section.Title>{__('Status')}</Section.Title>

        <SidebarList>
          {statusFilters.map((status, index) => (
            <li key={index}>
              <Link
                tabIndex={0}
                className={
                  router.getParam(history, 'status') === status.key
                    ? 'active'
                    : ''
                }
                to={`/engage?status=${status.key}`}
              >
                {__(status.value)}
                <SidebarCounter>{statusCounts[status.key]}</SidebarCounter>
              </Link>
            </li>
          ))}
        </SidebarList>
      </Section>
    );
  }

  render() {
    const { tags, tagCounts, config, toggleSection } = this.props;

    return (
      <Wrapper.Sidebar>
        <Box
          title={__('Kind')}
          name="showKinds"
          isOpen={config.showKinds || false}
          toggle={toggleSection}
        >
          {this.renderKindFilter()}
        </Box>
        <Box
          title={__('Status')}
          name="showStatus"
          isOpen={config.showStatus || false}
          toggle={toggleSection}
        >
          {this.renderStatusFilter()}
        </Box>
        <Box
          title={__('Filter by tags')}
          name="showFilter"
          isOpen={config.showFilter || false}
          toggle={toggleSection}
        >
          <CountsByTag
            tags={tags}
            manageUrl="tags/engageMessage"
            counts={tagCounts}
            loading={false}
          />
        </Box>
      </Wrapper.Sidebar>
    );
  }
}

export default Sidebar;
