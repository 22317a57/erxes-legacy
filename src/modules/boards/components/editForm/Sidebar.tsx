import Button from 'modules/common/components/Button';
import FormGroup from 'modules/common/components/form/Group';
import ControlLabel from 'modules/common/components/form/Label';
import CompanySection from 'modules/companies/components/common/CompanySection';
import { ICompany } from 'modules/companies/types';
import CustomerSection from 'modules/customers/components/common/CustomerSection';
import { ICustomer } from 'modules/customers/types';
import SelectTeamMembers from 'modules/settings/team/containers/SelectTeamMembers';
import PortableTasks from 'modules/tasks/components/PortableTasks';
import React from 'react';
import { Watch } from '../../containers/editForm/';
import { RightContent } from '../../styles/item';
import { IItem, IOptions } from '../../types';

type Props = {
  item: IItem;
  customers: ICustomer[];
  companies: ICompany[];
  assignedUserIds: string[];
  contentType?: string;
  onChangeField?: (
    name: 'companies' | 'customers' | 'assignedUserIds',
    value: any
  ) => void;
  copyItem: () => void;
  removeItem: (itemId: string) => void;
  sidebar?: () => React.ReactNode;
  extraContent?: React.ReactNode;
  options: IOptions;
};

class Sidebar extends React.Component<Props> {
  onChange = (type, value) => {
    const { onChangeField } = this.props;

    if (onChangeField) {
      onChangeField(type, value);
    }
  };

  renderAdditionalContent() {
    const {
      options,
      item,
      customers,
      companies,
      contentType,
      extraContent
    } = this.props;

    const cmpsChange = cmps => this.onChange('companies', cmps);
    const cmrsChange = cmrs => this.onChange('customers', cmrs);

    if (extraContent) {
      return extraContent;
    }

    return (
      <>
        <CompanySection
          name={options.title}
          companies={companies}
          onSelect={cmpsChange}
        />

        <CustomerSection
          name={options.title}
          customers={customers}
          onSelect={cmrsChange}
        />

        <PortableTasks contentType={contentType} contentId={item._id} />
      </>
    );
  }

  render() {
    const {
      item,
      copyItem,
      removeItem,
      sidebar,
      options,
      assignedUserIds
    } = this.props;

    const onClick = () => removeItem(item._id);
    const userOnChange = usrs => this.onChange('assignedUserIds', usrs);

    return (
      <RightContent>
        <FormGroup>
          <ControlLabel>Assigned to</ControlLabel>
          <SelectTeamMembers
            label="Choose users"
            name="assignedUserIds"
            value={assignedUserIds}
            onSelect={userOnChange}
            filterParams={{ status: 'verified' }}
          />
        </FormGroup>
        {sidebar && sidebar()}

        {this.renderAdditionalContent()}

        <Watch item={item} options={options} />

        <Button icon="checked-1" onClick={copyItem}>
          Copy
        </Button>

        <Button icon="cancel-1" onClick={onClick}>
          Delete
        </Button>
      </RightContent>
    );
  }
}

export default Sidebar;
