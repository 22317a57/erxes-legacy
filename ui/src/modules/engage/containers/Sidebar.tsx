import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import ErrorMsg from 'modules/common/components/ErrorMsg';
import Spinner from 'modules/common/components/Spinner';
import { IRouterProps } from 'modules/common/types';
import checkError from 'modules/common/utils/checkError';
import { queries as tagQueries } from 'modules/tags/graphql';
import { ITag } from 'modules/tags/types';
import React from 'react';
import { withRouter } from 'react-router';
import Sidebar from '../components/Sidebar';
import { queries } from '../graphql';
import { CountQueryResponse, TagCountQueryResponse } from '../types';

type Props = {
  queryParams: any;
} & IRouterProps;

const SidebarContainer = (props: Props) => {

  const { queryParams } = props;

  const {
    loading: kindCountsLoading,
    error: kindCountsError,
    data: kindCountsData
  } = useQuery<CountQueryResponse>(
    gql(queries.kindCounts));

  const {
    loading: statusCountsLoading,
    error: statusCountsError,
    data: statusCountsData
  } = useQuery<CountQueryResponse>(
    gql(queries.statusCounts), {
    variables: {
      kind: queryParams.kind || ''
    }
  });

  const {
    loading: tagsLoading,
    error: tagsError,
    data: tagsData
  } = useQuery<TagCountQueryResponse, { type: string }>(
    gql(tagQueries.tags), {
    variables: { type: 'engageMessage' }
  });

  const {
    loading: tagCountsLoading,
    error: tagCountsError,
    data: tagCountsData
  } = useQuery<CountQueryResponse, { kind: string; status: string }>(
    gql(queries.tagCounts), {
    variables: {
      kind: queryParams.kind || '',
      status: queryParams.status || ''
    }
  });

  if (kindCountsError || statusCountsError || tagsError || tagCountsError) {
    const error = checkError([kindCountsError, statusCountsError, tagsError, tagCountsError]);

    return <ErrorMsg>{error.message}</ErrorMsg>;
  };

  if (kindCountsLoading || statusCountsLoading || tagsLoading || tagCountsLoading) {
    return <Spinner objective={true} />;
  };


  const updatedProps = {
    ...props,
    kindCounts: (kindCountsData && kindCountsData.engageMessageCounts) || {},
    statusCounts: (statusCountsData && statusCountsData.engageMessageCounts) || {},
    tags: [tagsData && tagsData.tags] || [] as ITag[],
    tagCounts: (tagCountsData && tagCountsData.engageMessageCounts) || {}
  };

  return <Sidebar {...updatedProps} />;
};

export default withRouter<Props>(SidebarContainer)