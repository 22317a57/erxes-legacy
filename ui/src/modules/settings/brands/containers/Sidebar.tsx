import { useMutation, useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import ButtonMutate from 'modules/common/components/ButtonMutate';
import { IButtonMutateProps } from 'modules/common/types';
import { Alert, confirm } from 'modules/common/utils';
import React from 'react';
import { ChildProps } from 'react-apollo';
import Sidebar from '../components/Sidebar';
import { mutations, queries } from '../graphql';
import {
  BrandRemoveMutationResponse,
  BrandRemoveMutationVariables,
  BrandsCountQueryResponse,
  BrandsQueryResponse
} from '../types';

type Props = {
  queryParams: any;
  currentBrandId?: string;
  history?: any
};

const SidebarContainer = (props: ChildProps<Props>) => {
  const {
    queryParams,
    currentBrandId,
    history
  } = props;

  const {
    loading: brandsQueryLoading,
    error: brandsQueryError,
    data: brandsQueryData
  } = useQuery<BrandsQueryResponse, { perPage: number }>(
    gql(queries.brands),
    {
      variables: {
        perPage: queryParams.limit ? parseInt(queryParams.limit, 10) : 20
      },
      fetchPolicy: 'network-only'
    }
  );

  const {
    loading: brandsCountQueryLoading,
    error: brandsCountQueryError,
    data: brandsCountQueryData
  } = useQuery<BrandsCountQueryResponse>(gql(queries.brandsCount));

  const [removeMutation, { error: brandRemoveMutationError }] =
    useMutation<BrandRemoveMutationResponse, BrandRemoveMutationVariables>(
      gql(mutations.brandRemove), {
      refetchQueries: getRefetchQueries(queryParams, currentBrandId)
    });

  const brands = brandsQueryData ? brandsQueryData.brands : [];
  const brandsTotalCount = brandsCountQueryData ? brandsCountQueryData.brandsTotalCount : 0;

  // remove action
  const remove = brandId => {
    confirm().then(() => {
      removeMutation({
        variables: { _id: brandId }
      })
        .then(() => {
          Alert.success('You successfully deleted a brand.');
          history.push('/settings/brands');
        })
        .catch(error => {
          Alert.error(error.message);
        });
    });
  };

  const renderButton = ({
    name,
    values,
    isSubmitted,
    callback,
    object
  }: IButtonMutateProps) => {
    return (
      <ButtonMutate
        mutation={object ? mutations.brandEdit : mutations.brandAdd}
        variables={values}
        callback={callback}
        refetchQueries={getRefetchQueries(queryParams, currentBrandId)}
        isSubmitted={isSubmitted}
        type="submit"
        successMessage={`You successfully ${
          object ? 'updated' : 'added'
          } a ${name}`}
      />
    );
  };

  if (brandsQueryError || brandsCountQueryError || brandRemoveMutationError) {
    return <p>Error!</p>;
  }

  if (brandsCountQueryLoading || brandsQueryLoading) {
    return <p>Loading...</p>;
  }

  const updatedProps = {
    ...props,
    renderButton,
    brands,
    brandsTotalCount,
    remove,
    loading: brandsQueryLoading
  };

  return <Sidebar {...updatedProps} />;
};

const getRefetchQueries = (queryParams, currentBrandId?: string) => {
  return [
    {
      query: gql(queries.brands),
      variables: {
        perPage: queryParams.limit ? parseInt(queryParams.limit, 10) : 20
      }
    },
    {
      query: gql(queries.brands)
    },
    {
      query: gql(queries.integrationsCount)
    },
    {
      query: gql(queries.brandDetail),
      variables: { _id: currentBrandId || '' }
    },
    { query: gql(queries.brandsCount) }
  ];
};

export default SidebarContainer;
