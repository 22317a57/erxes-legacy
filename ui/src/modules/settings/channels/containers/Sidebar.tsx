import { useMutation, useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import ButtonMutate from 'modules/common/components/ButtonMutate';
import { IButtonMutateProps } from 'modules/common/types';
import { Alert, confirm } from 'modules/common/utils';
import React from 'react';
import Sidebar from '../components/Sidebar';
import { mutations, queries } from '../graphql';
import {
  ChannelsCountQueryResponse,
  ChannelsQueryResponse,
  RemoveChannelMutationResponse,
  RemoveChannelMutationVariables
} from '../types';

type Props = {
  queryParams: any;
  currentChannelId?: string;
  history?: any;
};

const SidebarContainer = (props: Props) => {
  const {
    queryParams,
    history,
    currentChannelId
  } = props;

  const {
    loading: channelsQueryLoading,
    error: channelsQueryError,
    data: channelsQueryData
  } = useQuery<ChannelsQueryResponse, { perPage: number }>(
    gql(queries.channels),
    {
      variables: {
        perPage: queryParams.limit ? parseInt(queryParams.limit, 10) : 20
      },
      fetchPolicy: 'network-only'
    }
  );

  const {
    loading: channelsCountQueryLoading,
    error: channelsCountQueryError,
    data: channelsCountQueryData
  } = useQuery<ChannelsCountQueryResponse>(gql(queries.channels));

  const [removeMutation, { error: channelRemoveMutationError }] =
    useMutation<RemoveChannelMutationResponse, RemoveChannelMutationVariables>(
      gql(mutations.channelRemove), {
      refetchQueries: getRefetchQueries(queryParams, currentChannelId)
    });

  // remove action
  const remove = channelId => {
    confirm().then(() => {
      removeMutation({
        variables: { _id: channelId }
      })
        .then(() => {
          Alert.success('You successfully deleted a channel.');

          history.push('/settings/channels');
        })
        .catch(error => {
          Alert.error(error.message);
        });
    });
  };

  const renderButton = ({
    name,
    values,
    isSubmitted,
    callback,
    object
  }: IButtonMutateProps) => {
    return (
      <ButtonMutate
        mutation={object ? mutations.channelEdit : mutations.channelAdd}
        variables={values}
        callback={callback}
        refetchQueries={getRefetchQueries(queryParams, currentChannelId)}
        isSubmitted={isSubmitted}
        type="submit"
        successMessage={`You successfully ${
          object ? 'updated' : 'added'
          } a ${name}`}
      />
    );
  };

  const channels = channelsQueryData ? channelsQueryData.channels : [];
  const channelsTotalCount = channelsCountQueryData ? channelsCountQueryData.channelsTotalCount : 0;

  if (channelsQueryError || channelsCountQueryError || channelRemoveMutationError) {
    return <p>Error!</p>;
  }

  if (channelsCountQueryLoading || channelsQueryLoading) {
    return <p>Loading...</p>;
  }

  const updatedProps = {
    ...props,
    channels,
    channelsTotalCount,
    remove,
    renderButton,
    loading: channelsQueryLoading
  };

  return <Sidebar {...updatedProps} />;
};

const getRefetchQueries = (queryParams, currentChannelId?: string) => {
  return [
    {
      query: gql(queries.channels),
      variables: {
        perPage: queryParams.limit ? parseInt(queryParams.limit, 10) : 20
      }
    },
    {
      query: gql(queries.channels),
      variables: {}
    },
    {
      query: gql(queries.integrationsCount),
      variables: {}
    },
    {
      query: gql(queries.channelDetail),
      variables: { _id: currentChannelId || '' }
    },
    { query: gql(queries.channelsCount) }
  ];
};

export default SidebarContainer;
