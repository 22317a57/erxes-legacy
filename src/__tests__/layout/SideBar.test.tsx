import { mount, shallow } from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';

import Sidebar from '../../modules/layout/components/Sidebar';

describe('Sidebar component', () => {
  const defaultProps = {
    children: 'child'
  };

  test('renders Sidebar successfully', () => {
    shallow(<Sidebar {...defaultProps} />);
  });

  test('renders successfully with default value', () => {
    const wrapper = mount(<Sidebar {...defaultProps} />);
    const props = wrapper.props();

    expect(props).toMatchObject(defaultProps);
  });

  test('snapshot matches', () => {
    const rendered = renderer.create(<Sidebar {...defaultProps} />).toJSON();

    expect(rendered).toMatchSnapshot();
  });
});
