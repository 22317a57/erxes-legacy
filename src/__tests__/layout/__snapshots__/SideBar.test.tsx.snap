// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component snapshot matches 1`] = `
<section
  className="sc-caSCKo gfAGes"
>
  <div
    className="sc-TOsTZ hEyXQp"
  >
    child
  </div>
</section>
`;
